src/gui/log/loglistview.cpp

```diff
@@ -0,0 +1,124 @@
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "loglistview.h"
+
+#include <QApplication>
+#include <QClipboard>
+#include <QFontMetrics>
+#include <QKeyEvent>
+#include <QPainter>
+#include <QStyle>
+#include <QStyledItemDelegate>
+
+#include "logmodel.h"
+#include "uithememanager.h"
+
+namespace
+{
+    const QString SEPERATOR = QStringLiteral(" - ");
+
+    class LogItemDelegate : public QStyledItemDelegate
+    {
+    public:
+        using QStyledItemDelegate::QStyledItemDelegate;
+
+    private:
+        void paint(QPainter *painter, const QStyleOptionViewItem &option, const QModelIndex &index) const override
+        {
+            const QPen originalPen = painter->pen();
+            painter->save();
+
+            QStyledItemDelegate::paint(painter, option, index); // paints background, focus rect and selection rect
+
+            const QStyle *style = option.widget ? option.widget->style() : QApplication::style();
+            const int textHMargin = style->pixelMetric(QStyle::PM_FocusFrameHMargin, &option, option.widget) + 1;
+            const int textVMargin = style->pixelMetric(QStyle::PM_FocusFrameVMargin, &option, option.widget);
+            const QRect textRect = style->subElementRect(QStyle::SE_ItemViewItemText, &option, option.widget)
+                                   .adjusted(textHMargin, textVMargin, -textHMargin, -textVMargin);
+
+            QPen coloredPen = originalPen;
+            coloredPen.setColor(Qt::gray);
+            painter->setPen(coloredPen);
+            const QString time = index.data(BaseLogModel::TimeRole).toString();
+            style->drawItemText(painter, textRect, 0, option.palette, (option.state & QStyle::State_Enabled), time);
+
+            painter->setPen(originalPen);
+            const int timeWidth = option.fontMetrics.width(time);
+            style->drawItemText(painter, textRect.adjusted(timeWidth, 0, 0, 0), 0, option.palette, (option.state & QStyle::State_Enabled), SEPERATOR);
+
+            const QVariant foreground = index.data(Qt::ForegroundRole);
+            if (foreground.canConvert<QColor>()) { // peermodel may give invalid foreground role
+                coloredPen.setColor(foreground.value<QColor>());
+                painter->setPen(coloredPen);
+            }
+
+            style->drawItemText(painter, textRect.adjusted((timeWidth + option.fontMetrics.width(SEPERATOR)), 0, 0, 0), 0
+                                , option.palette, (option.state & QStyle::State_Enabled), index.data(BaseLogModel::MessageRole).toString());
+            painter->restore();
+        }
+
+        QSize sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const override
+        {
+            const QString displayText = QString::fromLatin1("%1%2%3").arg(index.data(BaseLogModel::TimeRole).toString(), SEPERATOR
```

Comment:
Seems you can extract "build display text" logic into function to avoid code duplication.